src/algorithms/dqns/qr_dqn.jl

```diff
@@ -0,0 +1,158 @@
+    loss_func::Tf,
+    stack_size::Union{Int,Nothing} = nothing,
+    γ::Float32 = 0.99f0,
+    batch_size::Int = 32,
+    update_horizon::Int = 1,
+    min_replay_history::Int = 32,
+    update_freq::Int = 1,
+    ensemble_num::Int = 1,
+    target_update_freq::Int = 100,
+    traces = SARTS,
+    update_step = 0,
+    κ::Float32 = 1.0f0,
+    rng = Random.GLOBAL_RNG,
+) where {Tq,Tt,Tf}
+    copyto!(approximator, target_approximator)
+    sampler = NStepBatchSampler{traces}(;
+        γ = γ,
+        n = update_horizon,
+        stack_size = stack_size,
+        batch_size = batch_size,
+    )
+    N = ensemble_num
+    τ = collect(0:N+1)/N
+    τ  = (τ[2:N+1]+τ[1:N])/2
+    QRDQNLearner(
+        approximator,
+        target_approximator,
+        loss_func,
+        min_replay_history,
+        update_freq,
+        update_step,
+        target_update_freq,
+        sampler,
+        N,
+        rng,
+        κ,
+        τ,
+        0.0f0,
+    )
+end
+
+Flux.functor(x::QRDQNLearner) = (Q = x.approximator, Qₜ = x.target_approximator),
+y -> begin
+    x = @set x.approximator = y.Q
+    x = @set x.target_approximator = y.Qₜ
+    x
+end
+
+function (learner::QRDQNLearner)(env)
+    s = send_to_device(device(learner.approximator), state(env))
+    s = Flux.unsqueeze(s, ndims(s) + 1)
+    q = reshape(learner.approximator(s), :, learner.ensemble_num)
+    vec(mean(q, dims = 2)) |> send_to_host
+end
+
+function RLBase.update!(learner::QRDQNLearner, batch::NamedTuple)
+    Q = learner.approximator
+    Qₜ = learner.target_approximator
+    γ = learner.sampler.γ
+    loss_func = learner.loss_func
+    n = learner.sampler.n
+    batch_size = learner.sampler.batch_size
+    N = learner.ensemble_num
+    κ = learner.κ
+    τ = learner.τ
+    D = device(Q)
+
+    s, a, r, t, s′ = (send_to_device(D, batch[x]) for x in SARTS)
+    a = CartesianIndex.(repeat(batch.action, inner = N), 1:(N*batch_size))
+
+    target_q = reshape(Qₜ(s′), :, N, batch_size)
+    avg_q = mean(target_q, dims=2)
+
+    if haskey(batch, :next_legal_actions_mask)
+        l′ = send_to_device(D, batch[:next_legal_actions_mask])
+        avg_q .+= ifelse.(l′, 0.0f0, typemin(Float32))
+    end
+
+    aₜ = argmax(avg_q, dims=1)
+    aₜ = aₜ .+ typeof(aₜ)(CartesianIndices((0:0, 0:N-1, 0:0)))
+    qₜ = reshape(target_q[aₜ], :, batch_size)
+    target =
+    reshape(r, 1, batch_size) .+ γ * reshape(1 .- t, 1, batch_size) .* qₜ
+
+    gs = gradient(params(Q)) do
+        q = reshape(Q(s), :, N*batch_size)
+        q = q[a]
+
+        target = reshape(target, N, 1, batch_size)
+        q = reshape(q, 1, N, batch_size)
+
+        TD_error = (target .- q)
+        temp = Zygote.dropgrad(abs.(TD_error) .<  κ)
+        element_wise_huber_loss = ((TD_error.^2) .* temp) + κ*(TD_error .- 0.5*κ) .* (1 .- temp)
+
+        # dropgrad
+        element_wise_huber_loss =
+            abs.(reshape(τ, 1, N) .- Zygote.dropgrad(TD_error .< 0)) .*
+            element_wise_huber_loss ./ κ
```

Comment:
better to move the quntile huberloss part into a funtion

Reply:
But it is the same when implemented inline too right? But yea, can do that if it helps re-usability.

Reply:
At least for your current implementation, there's a slight difference in `Zygote.gradient` .

Reply:
Done.